kernel/src/arch/mipsel/board/malta/mod.rs

```diff
@@ -0,0 +1,54 @@
+use once::*;
+use alloc::string::String;
+use mips::registers::cp0;
+use crate::drivers::bus::pci;
+
+#[path = "../../../../drivers/serial/ti_16c550c.rs"]
+pub mod serial;
+#[path = "../../../../drivers/gpu/qemu_stdvga.rs"]
+pub mod vga;
+#[path = "../../../../drivers/gpu/fb.rs"]
+pub mod fb;
+#[path = "../../../../drivers/console/mod.rs"]
+pub mod console;
+pub mod consts;
+
+use fb::FramebufferInfo;
+
+/// Initialize serial port first
+pub fn init_serial_early() {
+    assert_has_not_been_called!("board::init must be called only once");
+    // initialize serial driver
+    serial::init(0xbf000900);
+    // Enable serial interrupt
+    unsafe {
+        let mut status = cp0::status::read();
+        status.enable_hard_int2();
+        cp0::status::write(status);
+    }
+    println!("Hello QEMU Malta!");
+}
+
+/// Initialize other board drivers
+pub fn init_driver() {
+    // TODO: add possibly more drivers
+    vga::init(0xbbe00000, 0xb2050000, 800, 600);
```

Comment:
can we find these address in dtb?

Reply:
NO. VGA is needed by framebuffer driver and video console, which initializes BEFORE the driver framework.

Also this is a board specific file, I don't see any problem with hardcoding addresses.

Reply:
Maybe a comprehensive refactor of the driver framework can solve this.